[YoutubeBridge] Fix bridge

csisoap opened the pull request:
- Add support for custom channel name (as request #2206).
- Add new method to fetch JSON data.
- Search, listing now parsing data through JSON, HTML is not work anymore.
- ~Remove page number input on the Search endpoint.~
- Merge two parseJSON function into one.
- ~Add ability to get better timestamp.~
- Fix regex on duration limit.
- Add anchor on some link in the description text.
- Add URL on the getURI() for each context.
**UPDATE 20/07/2021:**
- ~Add support for Youtube Data API v3.~ => Temporary remove API support, going to create a new PR after this PR merged.
- Fix playlists that have more than 15 items not display correctly.
- Add new custom name parameter.
- Fix sorting on playlist not function correctly if use API.
- Fix duration when try to get timestamp from Premiering video. (I haven't tested this on API)
- Update regex for URL.

Notes: This bridge will handle the case like Video, Livestream. Other videos like Premiering stuff not tested.

csisoap committed:
[YoutubeBridge] Fix bridge

- Add support for custom channel name.
- Add new method to fetch JSON data.
- Search, listing now parsing data through JSON, HTML is not work anymore.
- Remove page number input on the Search endpoint.

csisoap committed:
[YoutubeBridge] Fix getURI()

csisoap committed:
Revert "[YoutubeBridge] Fix getURI()"

This reverts commit 589c918d3be44092237cffcc22a872953cfcdadb.

csisoap committed:
[YoutubeBridge] Fix getURI()

csisoap committed:
[YoutubeBridge] Fix lint

csisoap committed:
[YoutubeBridge] Fix getURI(), fix regex

csisoap committed:
[YoutubeBridge] Fix edge case, add anchor to description

csisoap committed:
[YoutubeBridge] Fix lint

csisoap committed:
[YoutubeBridge] Add target=_blank for anchor tag

csisoap committed:
[YoutubeBridge] Fix YouTube subscribe URL being parsed, not redirect URL

csisoap committed:
[YoutubeBridge] Add better timestamp, fix typo

csisoap committed:
[YoutubeBridge] New change

- Merge two parseJSON into one.
- Remove ytBridgeParseHtmlParsing() due to YouTube not populating HTML in the first run.
- Add ability to get better timestamp.
- Handle the case of 'Livestream' video.

csisoap committed:
[YoutubeBridge] Add support for more URL type on Youtube

em92 commented on bridges/YoutubeBridge.php:
```diff
@@ -39,11 +39,6 @@ class YoutubeBridge extends BridgeAbstract {
-			'pa' => array(
-				'name' => 'page',
-				'type' => 'number',
-				'exampleValue' => 1
```
Deleting this parameter will make previous generated feed links invalid. As for now, I suggest to leave it as for now.

csisoap replied:
@em92 I've pushed new commit that fixed.

em92 reviewed (changes requested):
Empty feed if using search. https://feed.eugenemolotov.ru/pr2208/?action=display&bridge=Youtube&context=Search+result&s=Front+Mission+3&duration_min=&duration_max=&format=Html

Playlist stopped working.
Before PR: https://feed.eugenemolotov.ru/?action=display&bridge=Youtube&context=By+playlist+Id&p=PLJYf0JdTApCqAbZImkQagXEuByh-b_7To&duration_min=&duration_max=&format=Html
After PR: https://feed.eugenemolotov.ru/pr2208/?action=display&bridge=Youtube&context=By+playlist+Id&p=PLJYf0JdTApCqAbZImkQagXEuByh-b_7To&duration_min=&duration_max=&format=Html

csisoap committed:
[YoutubeBridge] New fix

- Fix condition error that make some feed stop working.
- Fix feed name in the search feed.
- Add back the page number parameter on the search feed.

em92 commented:
https://feed.eugenemolotov.ru/pr2208/?action=display&bridge=Youtube&context=By+playlist+Id&p=PLaDrN74SfdT7Ueqtwn_bXo1MuSWT0ji2w&duration_min=&duration_max=&format=Html

Item timestamps are incorrect. All videos show July 13 (today) with different years.

Previous version of timestamp fetching worked fine. Isn't it?

em92 commented on bridges/YoutubeBridge.php:
```diff
@@ -221,31 +202,164 @@ private function ytGetSimpleHTMLDOM($url, $cached = false){
+			$count++;
+			if($count == 20) {
+				break;
+			}
```
Playlist can have videos sorted by ascending (oldest first) upload date. This construction will ignore latest videos, which is not good.

For example: https://www.youtube.com/playlist?list=PLaDrN74SfdT7Ueqtwn_bXo1MuSWT0ji2w

csisoap commented:
> https://feed.eugenemolotov.ru/pr2208/?action=display&bridge=Youtube&context=By+playlist+Id&p=PLaDrN74SfdT7Ueqtwn_bXo1MuSWT0ji2w&duration_min=&duration_max=&format=Html
> 
> Item timestamps are incorrect. All videos show July 13 (today) with different years.
> 
> Previous version of timestamp fetching worked fine. Isn't it?

I looked into the video, for example like this: https://www.youtube.com/watch?v=UP091FeB-3o and see the timestamp is 2 years ago 15 minutes after extracted. So that why you see all videos show today with different years.

csisoap commented:
I'm going to remove those timestamp and set it back to use the meta timestamp instead.
The only way to get the corrected timestamp beside RSS feed is to using API from YouTube.

csisoap committed:
[YoutubeBridge] Remove codes extracted timestamp from accessibilityData, set timestamp back to those in meta tag.

csisoap committed:
[YoutubeBridge] Fix indent

csisoap committed:
[YoutubeBridge] Remove counter

em92 commented:
> Add support for custom channel name

This must be implemented as the other context, not as "User". For example, these channels are different:
https://www.youtube.com/c/DeathmatchTV
https://www.youtube.com/user/DeathmatchTV

Everything else looks good and ready to be merged once resolved issue above.

csisoap committed:
[YoutubeBridge] New change

- Add support for Youtube Data API v3.
- Fix playlists that have more than 15 items not display correctly.
- Add new custom name parameter.
- Fix sorting on playlist not function correctly if use API.

csisoap committed:
[YoutubeBridge] Fix lint

csisoap committed:
[YoutubeBridge] Fix lint

csisoap committed:
[YoutubeBridge] Remove unused variable

csisoap committed:
[YoutubeBridge] Add feed name for custom name, move some codebase to avoid repeat it

csisoap commented:
Temporary marking this as draft until PR #2100 is merged

em92 commented:
hi, @csisoap!
How is this PR related to that PR? That PR btw is blocked by other (droping support of php 5.6 and 7.0)

csisoap commented:
> hi, @csisoap!
> How is this PR related to that PR? That PR btw is blocked by other (droping support of php 5.6 and 7.0)

Hi @em92, this PR does supported Youtube Data API v3, which requires an API key. And that PR supported using environment variable for the bridge. I'm intent to add support for getting API key from environment variable once that PR is merged to avoid people accidently commited to the public repo.

em92 commented:
hi again, @csisoap!
Could you please temporary take out Youtube Data API v3 support and add new PR after I merge this PR? There are other some good features and fixes that you implemented. #2100 problably will be long to accept since RSS-Bridge needs to drop php 5.6 and 7.0.

After you take that API usage out, I will check PR again and, if everything works fine, I will accept it.

csisoap committed:
[YoutubeBridge] Remove support for YouTube Data API v3

csisoap commented:
You can merge it now, @em92

em92 reviewed (changes requested):
Following links return empty feed:
https://feed.eugenemolotov.ru/pr2208/?action=display&bridge=Youtube&context=Search+result&s=Front+Mission+3&duration_min=&duration_max=&format=Html (search)
https://feed.eugenemolotov.ru/pr2208/?action=display&bridge=Youtube&context=By+playlist+Id&p=PLaDrN74SfdT7Ueqtwn_bXo1MuSWT0ji2w&duration_min=&duration_max=&format=Html (playlist)

Current version returns non-empty playlist: https://feed.eugenemolotov.ru/?action=display&bridge=Youtube&context=By+playlist+Id&p=PLaDrN74SfdT7Ueqtwn_bXo1MuSWT0ji2w&duration_min=&duration_max=&format=Html

csisoap committed:
[YoutubeBridge] Fix no results were displayed after scraping.

csisoap commented:
> Following links return empty feed:
> https://feed.eugenemolotov.ru/pr2208/?action=display&bridge=Youtube&context=Search+result&s=Front+Mission+3&duration_min=&duration_max=&format=Html (search)
> https://feed.eugenemolotov.ru/pr2208/?action=display&bridge=Youtube&context=By+playlist+Id&p=PLaDrN74SfdT7Ueqtwn_bXo1MuSWT0ji2w&duration_min=&duration_max=&format=Html (playlist)
> 
> Current version returns non-empty playlist: https://feed.eugenemolotov.ru/?action=display&bridge=Youtube&context=By+playlist+Id&p=PLaDrN74SfdT7Ueqtwn_bXo1MuSWT0ji2w&duration_min=&duration_max=&format=Html

Oops, my bad. Sorry about that. I've pushed a commit that resolve this.

em92 commented:
gj!
